frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/model/ResourceDescriptor.java

```diff
@@ -130,82 +190,156 @@ public static ResourceDescriptor fromResource(Resource res) throws Exception {
+        LOGGER.logError(e, "Ignore: Failed to set GPU information, illegal access function");
       }
     }
-    return res;
-  }
 
-  @Override
-  public String toString() {
-    return String.format("[MemoryMB: [%s]", getMemoryMB()) + " " +
-        String.format("CpuNumber: [%s]", getCpuNumber()) + " " +
-        String.format("GpuNumber: [%s]", getGpuNumber()) + " " +
-        String.format("GpuAttribute: [%s]]", CommonExts.toStringWithBits(getGpuAttribute()));
+    if (portRanges != null && portRanges.size() > 0) {
+      try {
+
+        Class hadoopValueRangesClass = Class.forName("org.apache.hadoop.yarn.api.records.ValueRanges");
+        Class hadoopValueRangeClass = Class.forName("org.apache.hadoop.yarn.api.records.ValueRange");
+
+        Object obj = hadoopValueRangesClass.getMethod("newInstance").invoke(null);
+
+        List<Object> hadoopValueRangeList = new ArrayList<Object>();
+
+        for (ValueRange range : portRanges) {
+          Object valueRangeObj = hadoopValueRangeClass.getMethod("newInstance", int.class, int.class).invoke(null, range.getBegin(), range.getEnd());
+          hadoopValueRangeList.add(valueRangeObj);
+        }
+
+        Method setRangesList = hadoopValueRangesClass.getMethod("setRangesList", List.class);
+        setRangesList.invoke(obj, hadoopValueRangeList);
+
+        Method setPort = clazz.getMethod("setPorts", hadoopValueRangesClass);
+        setPort.invoke(res, obj);
+
+      } catch (NoSuchMethodException e) {
+        LOGGER.logDebug(e, "Ignore: Failed to get Ports information, YARN library doesn't support Port");
+      } catch (IllegalAccessException e) {
+        LOGGER.logError(e, "Ignore: Failed to get Ports information, illegal access function");
+      } catch (ClassNotFoundException e) {
+        LOGGER.logDebug(e, "Ignore: Failed to get the class Name");
+      }
+    }
+    LOGGER.logDebug("Put LocalResource " + this.toString() + " to hadoop resource: " + res);
+    return res;
   }
 
   // Maybe underestimate if any GpuAttribute == 0
   public static ResourceDescriptor subtract(ResourceDescriptor lhs, ResourceDescriptor rhs) {
-    ResourceDescriptor ret = new ResourceDescriptor();
-    ret.setMemoryMB(lhs.getMemoryMB() - rhs.getMemoryMB());
-    ret.setCpuNumber(lhs.getCpuNumber() - rhs.getCpuNumber());
-    ret.setGpuAttribute(lhs.getGpuAttribute() & (~(rhs.getGpuAttribute())));
-    if (lhs.getGpuAttribute() != 0 && rhs.getGpuAttribute() != 0) {
-      ret.setGpuNumber(Long.bitCount(ret.getGpuAttribute()));
-    } else {
-      ret.setGpuNumber(lhs.getGpuNumber() - rhs.getGpuNumber());
-    }
+    ResourceDescriptor ret = YamlUtils.deepCopy(lhs, ResourceDescriptor.class);
+    ResourceDescriptor.subtractFrom(ret, rhs);
     return ret;
   }
 
-  // Maybe overestimate if any GpuAttribute == 0
+  public static ResourceDescriptor subtractFrom(ResourceDescriptor lhs, ResourceDescriptor rhs) {
```

Comment:
>subtractFrom [](start = 35, length = 12)

Can we remove this function? We do not want to change input always.

Reply:
No need deepcopy, please revert to the original way for this method,

---
In reply to: [174759653](https://github.com/Microsoft/pai/pull/318#discussion_r174759653) [](ancestors = 174759653)

Reply:
please revert to the original way for this method,

---
In reply to: [175414127](https://github.com/Microsoft/pai/pull/318#discussion_r175414127) [](ancestors = 175414127,174759653)

Reply:
if (lhs.getGpuAttribute() != 0 && rhs.getGpuAttribute() != 0) {
      ret.setGpuNumber(Long.bitCount(ret.getGpuAttribute()));
    } else {
      ret.setGpuNumber(lhs.getGpuNumber() - rhs.getGpuNumber());
    }

---
In reply to: [175990731](https://github.com/Microsoft/pai/pull/318#discussion_r175990731) [](ancestors = 175990731,175414127,174759653)

Reply:
if (lhs.getGpuAttribute() != 0 && rhs.getGpuAttribute() != 0) {
      ret.setGpuNumber(Long.bitCount(ret.getGpuAttribute()));
    } else {
      ret.setGpuNumber(lhs.getGpuNumber() - rhs.getGpuNumber());
    }

---
In reply to: [175990731](https://github.com/Microsoft/pai/pull/318#discussion_r175990731) [](ancestors = 175990731,175414127,174759653)